### DISCUSSION TITLE
RDF4J on Debian 10: Logging problems

### DISCUSSION THREAD

**Discussion by volkerjaenisch** [eclipse-rdf4j/rdf4j#3133] [Q&A] [answered]
Dear RDF4J People!

The last two days I struggled to install RDF4J on a plain new Debian 10 system.

The RDF4J-Server component I run in Tomcat as RAW with the following result.

`Invocation of init method failed; nested exception is java.io.IOException: Unable to create logging directory /.RDF4J/server/logs
`
Setup:
From debian packages:
* openjdk-11
* tomcat9
* libslf4j-java
* liblog4j1.2-java

RDF4J from the current tar.gz install. WAR files copied to /var/lib/tomcat/webapps/.

I think this is a not too exotic setup.

Running the same WAR in jetty I get the same exception but with a bit more friendly path:

`Invocation of init method failed; nested exception is java.io.IOException: Unable to create logging directory /.RDF4J/server/logs`

When I run the code in eclipse (the console application) the same logging code produces /home/<user>/.RDF4J/server/logs

Any cases are for from optimal. No one likes to give write a web app access to paths like that :-).

The strace's I did show me that RDF4J does look for logging config files but at weird positions e.g.

```
/var/lib/tomcat9/webapps/rdf4j-server/WEB-INF/classes/org/eclipse/rdf4j/common/app/config/defaults/logging.properties
/var/lib/tomcat9/lib/org/eclipse/rdf4j/common/app/config/logging.properties
```

But RDF4J never ever ask for a logback.xml file which is given by the WAR.

Surely I miss something here.

Any help appreciated,

Volker

**Comment by barthanssens** [comment, marked as answer]
Hi Volker,

have you tried setting the `org.eclipse.rdf4j.appdata.basedir`  (and if you're running tomcat as a systemd service, you might also need to modify the service file) ? That might do the trick...

See also
* https://rdf4j.org/documentation/tools/server-workbench/#application-directory-configuration
* https://rdf4j.org/documentation/tools/server-workbench/#logging-configuration

**Comment by volkerjaenisch** [comment]
@barthanssens 

Thank you so much for the fast answer. I have solved the problem in no time.

I really missed that documentation bit. Installation is a quite complex case and if you are not familiar with Java and the where abouts you can get lost quickly. 

Same with Python installation in principal. If you come from Java you will be lost at first.

But : In the python plane there are for any application some "walk through" tutorials for some platforms. These are not complete, nor accurate, nor perfect but they give you a base line, something to work with. I really miss such a thing in the java plane.

I did not found any base line tutorial to setup RDF4J for non Java Programmers.
The RDF4J Documentation may be complete in a mathematical sense for the JAVA programmer, but for the beginners they are of no use.  

I have the experience of 40 years of programing and I stuck. 

Cheers,
Volker

**Reply by abrokenjester** [reply]
@volkerjaenisch fwiw we do have some beginner's tutorials on the site, but as you say, they are more about how to get started with RDF4J as a programming library toolkit, rather than as a standalone tool. It's an aspect of the project that has been a bit neglected in recent years, I'm afraid to say. I saw further down that you are writing a tutorial/howto that is more aimed at this aspect, if you do decide to contribute that at some point, we'd be overjoyed to have it - or if you prefer to publish under your own name just ping us and I'm sure we can find a spot to link to it.

**Comment by barthanssens** [comment]
Hi Volker,

glad it worked that way... admittedly the getting started part of the documentation can be improved, most of the documentation is indeed written for/by Java developers.
FWIW, we're also working on a more easy way to set up a read-only endpoint (#2904) and integration with Spring framework (#3059)

**Comment by volkerjaenisch** [comment]
@barthanssens 
Thanks for the help you offered, so quickly and to the point.

We are in the uncomfortable situation that we utilized  Blazegraph and it is dead now. There is also a 10 year and running  "constriction" on the RDF Graph-DB-Market, driven by the "neuronal nets can do it all hype". 
So I suppose we are not the only "non Java" developers who look for an alternative Triple DB, at the moment.

- Neo4j installs and runs in 20 seconds. 

- Apache Jena is a pain with a crappy documentation. 
- RDF4J is installed quite easily. But does not come up since a logging-path points into Nirvana. And you need two days to ensure that you are not a complete idiot and that you need external help.

There is a pattern, you will noticed :-)
Neo4J can charge their customers badly since there is no real "get the things done" alternative. On the other side they sell RDF4-Code - which is completely legal - but has a smell.

I like your documentation. It is way better than that of Jena. Also I did not get the last 3 versions of Jena to install via maven.

So I decided to base my trust on RDF4J. I was rewarded by your kind support and will stay here. I also will contribute as an offering to the goddess of good fortune an "Howto RDF4J for Python developers" on my company website. 
If this Howto is mature it may be one day become a part of the RDF4J documentation.

Cheers,
Volker
